Implement Covering Arrays

aadwyer opened the pull request:
Fixes #34279

aadwyer committed:
This is the initial commit, the covering array class was created
with a few initial methods: is_covering_array and the init method,
as well as that give its size n and k, its parameters t and v,
and an ordering based on the lexicographic ordering of each row.

aadwyer committed:
This is the initial commit, the covering array class was created
with a few initial methods: is_covering_array and the init method,
as well as that give its size n and k, its parameters t and v,
and an ordering based on the lexicographic ordering of each row.

aadwyer committed:
Created py file for basic CA generation methods, will not be used in first
push so put it in new branch

aadwyer committed:
added comments to covering array file, still to add examples

aadwyer committed:
finished the strength 2 construction

aadwyer committed:
Added some methods namely pprint and __hash, edited/added doctests some of
which are not yet working

aadwyer committed:
Ensured all the doctests are working.

aadwyer committed:
removed covering array file temporarily

aadwyer committed:
Merge branch 'u/gh-Aaron-Dwyer/coveringarrays' of trac.sagemath.org:sage into t/34279/coveringarrays

aadwyer committed:
The covering arrays file is now up to date

aadwyer committed:
Fixed the merge of the covering array file

aadwyer committed:
Moved the is_covering_array function to be a class method. Allowed
use of specific symbol sets for CAs  instead of just their size.

aadwyer committed:
Renamed file

aadwyer committed:
added methods

dimpase committed:
Merge branch 'develop' into t/34279/coveringarrays

aadwyer committed:
Fixed the name of the file in the doctests, and deleted a temporary file

aadwyer committed:
Merge branch '#35008' into t/34279/coveringarrays

aadwyer committed:
Merge branch 'develop' into t/34279/coveringarrays

aadwyer committed:
Merge branch 'develop' into t/34279/coveringarrays

aadwyer committed:
Edited file to fix unsuccessful checks

aadwyer committed:
Merge branch 't/34279/coveringarrays' of github.com:Aaron-Dwyer/sage into t/34279/coveringarrays

aadwyer committed:
Merge branch 'develop' into t/34279/coveringarrays

brettpim committed:
added two doctests:
- in .symbolset if we don't specify symbols set on init
- in .strength if we pass it strength 3 CA

aadwyer committed:
Merge pull request #1 from brettpim/covering_array_two

added two doctests:

aadwyer committed:
Merge branch 'develop' into t/34279/coveringarrays

dimpase reviewed (commented):
I'll look at it tomorrow

dimpase commented on src/sage/combinat/designs/covering_array.py:
```diff
@@ -0,0 +1,760 @@
+
+    def strength(self):
+        r"""
+        Return the strength of the covering array, which is the paramter
```
typo - paramter

dimpase committed:
typo fixes

dimpase reviewed (changes requested):
You need to hook this up to documentation, its indices, etc. E.g. have a look how it's don't for covering designs. (Please shout if you need more details on this).
Currently `CoveringArray` is nowhere to be found, it seems.

aadwyer committed:
Added the covering array file to documentation, fixed a couple formatting
typos in the covering array file

aadwyer committed:
Merge pull request #3 from Aaron-Dwyer/covering_array

Added the covering array file to documentation, fixed a couple format…

aadwyer committed:
Update covering_array.py

aadwyer committed:
added covering array file to documentation, and made some changes
to the file including richcmp and some formatting changes

aadwyer committed:
Fixed some more formatting issues and doctests for CA file, richcmp
is not yet working

aadwyer committed:
Merge branch 't/34279/coveringarrays' into covering_array

aadwyer committed:
Merge pull request #4 from Aaron-Dwyer/covering_array

Covering array

aadwyer committed:
Merge branch 'develop' into t/34279/coveringarrays

aadwyer committed:
added decorator to richcmp method as needed, linked references

aadwyer committed:
Added covering arrays to designs/design_catalog.py. Also edited CA
file, changing some variable names and some doctest formatting,
also sset is changed to a tuple and changed functions to use
internal attributes

aadwyer committed:
Merge branch 'covering_array' of github.com:Aaron-Dwyer/sage into covering_array

aadwyer committed:
Merge pull request #5 from Aaron-Dwyer/covering_array

Covering array

aadwyer committed:
Removed covering array file, instead put is_covering_array in
designs_pyx to be consistent with is_orthogonal_array.

Since covering array file is gone, removed links to it in module
list and design catalogue.

aadwyer committed:
Fixed some typos

aadwyer committed:
fixed another typo

aadwyer committed:
fixing errors

aadwyer committed:
fixed another typo again

aadwyer committed:
fixed another typo again

aadwyer committed:
Removed use of "copy" and condensed multiple lines to one

aadwyer committed:
Merge pull request #7 from Aaron-Dwyer/coveringarray

Coveringarray

aadwyer committed:
Merge pull request #8 from Aaron-Dwyer/coveringarray

fixed another typo

aadwyer committed:
Merge pull request #9 from Aaron-Dwyer/coveringarray

fixing errors

videlec commented on src/sage/combinat/designs/designs_pyx.pyx:
```diff
@@ -15,6 +15,203 @@ from cysignals.memory cimport sig_malloc, sig_calloc, sig_realloc, sig_free
+        tupledict={}
+        a=[ttuple for ttuple in product(symbol_set,repeat=wstrength)]
+        for item in a:
+            tupledict.update({item:0})
```
I do not understand the four lines above. Aren't they equivalent to the one line
```python
tupledict = {item: 0 for item in product(symbol_set, repeat=wstrength)}
```

videlec commented on src/sage/combinat/designs/designs_pyx.pyx:
```diff
@@ -15,6 +15,203 @@ from cysignals.memory cimport sig_malloc, sig_calloc, sig_realloc, sig_free
+        # Iterates over every possible selections of t columns, and
+        #count the t-tuples appearing in each selection
+        for comb in combinations(range(number_columns),wstrength):
+            wdict=deepcopy(tupledict)
```
I would suggest to not use `deepcopy` but instead use the constructor I suggested in my first comment. Namely
```python
wdict = {item: 0 for item in product(symbol_set, repeat=wstrength)}
```
That way you can simply avoid creating `tupledict` at all.

aadwyer committed:
Merge pull request #10 from Aaron-Dwyer/coveringarray

Removed use of "copy" and condensed multiple lines to one

videlec commented on src/sage/combinat/designs/designs_pyx.pyx:
```diff
@@ -15,6 +15,195 @@ from cysignals.memory cimport sig_malloc, sig_calloc, sig_realloc, sig_free
+    - ``array`` -- The Covering Array to be tested.
+
+    - ``strength`` (integer) -- The parameter `t` of the covering array,
+      such thatin any selection of `t` columns of the array, every `t`
```
thatin -> thatin

videlec commented on src/sage/combinat/designs/designs_pyx.pyx:
```diff
@@ -15,6 +15,195 @@ from cysignals.memory cimport sig_malloc, sig_calloc, sig_realloc, sig_free
+
+    - ``strength`` (integer) -- The parameter `t` of the covering array,
+      such thatin any selection of `t` columns of the array, every `t`
+      -tuple appearsat least once. If set to None then all t>0 are
```
appearsat -> appears at

videlec commented on src/sage/combinat/designs/designs_pyx.pyx:
```diff
@@ -15,6 +15,195 @@ from cysignals.memory cimport sig_malloc, sig_calloc, sig_realloc, sig_free
+    - ``strength`` (integer) -- The parameter `t` of the covering array,
+      such thatin any selection of `t` columns of the array, every `t`
+      -tuple appearsat least once. If set to None then all t>0 are
+      tested to and themaximal strength is used.
```
themaximal -> the maximal

videlec commented on src/sage/combinat/designs/designs_pyx.pyx:
```diff
@@ -15,6 +15,195 @@ from cysignals.memory cimport sig_malloc, sig_calloc, sig_realloc, sig_free
 
+def is_covering_array(array,strength=None,symbol_set=None,verbose=False,parameters=False):
+    r"""
+    Checks if the input is a covering array with given strength.
```
Checks -> Check (see [developer guide](https://doc.sagemath.org/html/en/developer/coding_basics.html#the-docstring-of-a-function-content))

videlec commented on src/sage/combinat/designs/designs_pyx.pyx:
```diff
@@ -15,6 +15,195 @@ from cysignals.memory cimport sig_malloc, sig_calloc, sig_realloc, sig_free
+      the Covering Array. If set to ``True``, the function returns a
+      pair ``(boolean_answer,(N,t,k,v))``.
+
+    EXAMPLES::
```
There must be a line break after `EXAMPLES::`

videlec commented on src/sage/combinat/designs/designs_pyx.pyx:
```diff
@@ -15,6 +15,195 @@ from cysignals.memory cimport sig_malloc, sig_calloc, sig_realloc, sig_free
 
 from sage.misc.unknown import Unknown
 
+def is_covering_array(array,strength=None,symbol_set=None,verbose=False,parameters=False):
```
There must be spaces after `,`, namely
```
def is_covering_array(array, strength=None, symbol_set=None, verbose=False, parameters=False):
```

aadwyer committed:
Fixed and error with the main loop of is_covering_array that was
terminating too early. Changed the function to only consider
arrays with symbols 0...v for some input v. Changed doctests and
sanity checks accordingly.

aadwyer committed:
For efficiency made is_covering_array only consider the arrays where
the symbols are 0,..,v-1. To allow for easy transitions between this
"standard symbol set" and others, edits were made to the orthogonal
arrays file to include functions that map symbol sets of arrays and
these functions are imported to the file for covering array functions

Because the covering array file was added back, documentation and
links to it were added back as well

aadwyer committed:
Merge pull request #11 from Aaron-Dwyer/coveringarray

Coveringarray

videlec commented on src/sage/combinat/designs/orthogonal_arrays.py:
```diff
@@ -1543,8 +1553,42 @@ def OA_relabel(OA,k,n,blocks=tuple(),matrix=None):
+    """
+    symbol_list = sorted({x for l in OA for x in l})
+    result = []
+    the_map = lambda x: symbol_list.index(x)
```
This takes quadratic time for no good reason. You should rather use a dictionary
```
mapping = {symbol : index for index, symbol in enumerate(symbol_list)}
```
However, that would fail for non hashable symbols. Nevertheless you also assumed the symbols to be hashable in your first line.

tscrim replied:
+1 on this change; almost everything people will use will be hashable (or at least should be and just needs the immutable version).

aadwyer replied:
I improved both functions that use this line with your suggestion

aadwyer committed:
Made the new relabelling functions in OA more efficient, removed an
import of the CA class which no longer exists

aadwyer committed:
Merge pull request #12 from Aaron-Dwyer/coveringarray

Made the new relabelling functions in OA more efficient, removed an import of the CA class which no longer exists

aadwyer committed:
Merge branch 'develop' into t/34279/coveringarrays

aadwyer committed:
Merge branch 'develop' into t/34279/coveringarrays

videlec commented on src/sage/combinat/designs/designs_pyx.pyx:
```diff
@@ -15,6 +15,202 @@ from cysignals.memory cimport sig_malloc, sig_calloc, sig_realloc, sig_free
+    result = True
+    # If no strength inputted, try increasing values for t until one
+    # does not work. If strength is inputted end after one check
+    while finished is False:
```
replace by `while not finished`

videlec commented on src/sage/combinat/designs/orthogonal_arrays.py:
```diff
@@ -1514,6 +1520,10 @@ def OA_relabel(OA,k,n,blocks=tuple(),matrix=None):
         True
 
+        sage: OA = designs.orthogonal_arrays.build(3,2)
+        sage: OA_relabel(OA,3,2,symbol_list=['A','B'])
```
Please add a space after each comma, ie `OA_relabel(OA, 3, 2, symbol_list=['A', 'B']))` rather than `OA_relabel(OA,3,2,symbol_list=['A','B'])`

videlec commented on src/sage/combinat/designs/orthogonal_arrays.py:
```diff
@@ -1543,8 +1553,36 @@ def OA_relabel(OA,k,n,blocks=tuple(),matrix=None):
+    """
+    symbol_list = sorted({x for l in OA for x in l})
+    mapping = {symbol : index for index, symbol in enumerate(symbol_list)}
+    return [[mapping.get(element) for element in row] for row in OA]
```
idem 1558

videlec commented on src/sage/combinat/designs/orthogonal_arrays.py:
```diff
@@ -1543,8 +1553,36 @@ def OA_relabel(OA,k,n,blocks=tuple(),matrix=None):
         OA = [[matrix[i][j] if j is not None else None for i,j in enumerate(R)] for R in OA]
 
+    if symbol_list:
+        mapping = {index : symbol for index, symbol in enumerate(symbol_list)}
```
Please, no space before a colon `:`

videlec commented on src/sage/combinat/designs/orthogonal_arrays.py:
```diff
@@ -1543,8 +1553,36 @@ def OA_relabel(OA,k,n,blocks=tuple(),matrix=None):
 
+    if symbol_list:
+        mapping = {index : symbol for index, symbol in enumerate(symbol_list)}
+        OA = [[mapping.get(element) for element in row] for row in OA]
```
Rather than mapping.get(element) use mapping[element] which is the standard way of accessing element in a dictionary. It is one the one hand faster and also behaves differently when the element is not in the dictionary .

videlec commented on src/sage/combinat/designs/designs_pyx.pyx:
```diff
@@ -15,6 +15,202 @@ from cysignals.memory cimport sig_malloc, sig_calloc, sig_realloc, sig_free
+        # Iterate over every possible selections of t columns, and
+        # count the t-tuples appearing in each selection
+        for comb in combinations(range(number_columns), wstrength):
+            tuple_dictionary = {item: 0 for item in product(symbol_list, repeat=wstrength)}
```
Why do you need to count here? You just need to know that each combination appears, don't you? I would just create the set of row-combinations and check its size
```
existing_combinations = set(tuple([row[ti] for ti in comb]) for row in array)
if len(existing_combinations) != number_columns ** wstrength:
    ...
```

aadwyer committed:
Fixed some formatting and improved the main loop of is_covering_array

aadwyer committed:
Merge pull request #14 from Aaron-Dwyer/covering_array

Fixed some formatting and improved the main loop of is_covering_array

videlec commented on src/doc/en/reference/references/index.rst:
```diff
@@ -5555,6 +5558,10 @@ REFERENCES:
              Course material, 1997,
              `<http://www.cs.tau.ac.il/~rshamir/atga/atga.html>`_
 
+.. [Sher2006] G.B. Sherwood, S.S Martirosyan, and C.J. Colbourn, "Covering
```
For multiple authors, the rest of the file uses the concatenation of surnames first letters. To conform to the style of the file I would choose `[SMC2006]`.

Do you know why is it not in alphabetic order here ?

aadwyer replied:
for the references I listed the authors in the order appearing in the paper, should I order them alphabetically instead?

videlec replied:
I think it is preferable as you did. I was just curious (because very non-standard).

videlec commented on src/doc/en/reference/references/index.rst:
```diff
@@ -6265,6 +6272,10 @@ REFERENCES:
              Math. Combin. Comput. **36** (2001), 95-118.
              http://www.info2.uqam.ca/~walsh_t/papers/Involutions%20paper.pdf
 
+.. [Wal2007] R.A. Walker II, and C.J. Colbourn, "Perfect Hash Families:
```
Similar, I would use `[WC2007]`.

videlec commented on src/sage/combinat/designs/covering_array.py:
```diff
@@ -0,0 +1,43 @@
+
+- Aaron Dwyer and brett stevens (2022): initial version
+
+.. TODO::
```
This TODO does not add much to the content. If you want to keep a list of TODO things open an issue on github with possibly explicit tasks.

videlec commented on src/sage/combinat/designs/covering_array.py:
```diff
@@ -0,0 +1,43 @@
+
+A Covering Array, denoted CA`(N,k,v,t)`, is an `n` by `k` array with
+entries from a set of `v` elements with the property that in every
+selection of `t` columns, eachrow contains every sequence of
```
`eachrow` -> `each row`

videlec commented on src/sage/combinat/designs/covering_array.py:
```diff
@@ -0,0 +1,43 @@
+
+An Orthogonal Array, denoted OA`(N,k,v,t)` is a covering array with the
+property that each row contains every sequence of `t`-elements exactly
+once
```
Missing a dot at the end of the sentence. Also you should add a link to the relevant module dealing with orthogonal arrays here.

videlec commented on src/sage/combinat/designs/covering_array.py:
```diff
@@ -0,0 +1,43 @@
+
+    Implement various construction methods for CAs
+
+"""
```
It feels like the module documentation is lacking examples. You should have a look at `orthogonal_arrays.py` for inspiration.

videlec commented on src/sage/combinat/designs/designs_pyx.pyx:
```diff
@@ -15,6 +15,199 @@ from cysignals.memory cimport sig_malloc, sig_calloc, sig_realloc, sig_free
 
 from sage.misc.unknown import Unknown
 
+#Current Version
```
What is this comment doing here?

videlec commented on src/sage/combinat/designs/designs_pyx.pyx:
```diff
@@ -15,6 +15,199 @@ from cysignals.memory cimport sig_malloc, sig_calloc, sig_realloc, sig_free
+    r"""
+    Check if the input is a covering array with given strength.
+
+    - ``array`` -- the Covering Array to be tested.
```
The list of arguments should be preceded with `INPUT:`, see [sage developer guide, docstring section](https://doc.sagemath.org/html/en/developer/coding_basics.html#the-docstring-of-a-function-content).

videlec commented on src/sage/combinat/designs/designs_pyx.pyx:
```diff
@@ -15,6 +15,199 @@ from cysignals.memory cimport sig_malloc, sig_calloc, sig_realloc, sig_free
+        symbol_list = list({x for l in array for x in l})
+        levels = len(symbol_list)
+    else:
+        symbol_list = [num for num in range(levels)]
```
Why not `symbol_list = range(levels)` ?

videlec commented on src/sage/combinat/designs/designs_pyx.pyx:
```diff
@@ -15,6 +15,199 @@ from cysignals.memory cimport sig_malloc, sig_calloc, sig_realloc, sig_free
+                break
+
+    if verbose:
+            print('A {} by {} Covering Array with strength {} with entries from a symbol set of size {}'.format(number_rows,number_columns,wstrength,levels))
```
Missing spaces after commas.

videlec commented on src/sage/combinat/designs/orthogonal_arrays.py:
```diff
@@ -1543,8 +1553,36 @@ def OA_relabel(OA,k,n,blocks=tuple(),matrix=None):
+        mapping = {index: symbol for index, symbol in enumerate(symbol_list)}
+        OA = [[mapping[element] for element in row] for row in OA]
     return OA
 
```
There should be two blank lines between function definitions in the module namespace, see [PEP8, section blank lines](https://peps.python.org/pep-0008/#blank-lines).

videlec commented on src/sage/combinat/designs/orthogonal_arrays.py:
```diff
@@ -1543,8 +1553,36 @@ def OA_relabel(OA,k,n,blocks=tuple(),matrix=None):
+    symbol_list = sorted({x for l in OA for x in l})
+    mapping = {symbol: index for index, symbol in enumerate(symbol_list)}
+    return [[mapping[element] for element in row] for row in OA]
+
```
idem : two blank lines

videlec reviewed (commented):
Some final comments.

videlec commented on src/sage/combinat/designs/covering_array.py:
```diff
@@ -0,0 +1,43 @@
+r"""
+Covering Arrays (CA)
+
+A Covering Array, denoted CA`(N,k,v,t)`, is an `n` by `k` array with
```
This definition is weirdly formulated : each row is unlikely to contain every sequence of `t`-elements.

aadwyer committed:
Fixed some typos and formatting errors, fixed a doctest

aadwyer committed:
fixed typo

aadwyer committed:
Merge pull request #15 from Aaron-Dwyer/covering_array

Covering array

aadwyer committed:
Added to the documentation of covering_array.py and split the
loop of the is_CA function into cases where t is inputted and
is not, so to have better iteration and better stopping

aadwyer committed:
Merge branch 'sagemath:develop' into covering_array

aadwyer committed:
Merge branch 'sagemath:develop' into covering_array

aadwyer committed:
Merge pull request #16 from Aaron-Dwyer/covering_array

Covering array

aadwyer committed:
Fixed formatting in documentaton

aadwyer committed:
Merge branch 'sagemath:develop' into covering_array

aadwyer committed:
Merge branch 't/34279/coveringarrays' into covering_array

aadwyer committed:
Merge pull request #19 from Aaron-Dwyer/covering_array

Covering array

aadwyer committed:
Merge branch 'develop' into t/34279/coveringarrays

dimpase committed:
Merge branch 'develop' into t/34279/coveringarrays
